src/dos/dos_mscdex.cpp

```diff
@@ -1072,10 +1072,24 @@ static uint16_t MSCDEX_IOCTL_Input(PhysPt buffer,uint8_t drive_unit) {
 				   };
 		case 0x0E :{ /* Get UPC */	
 					uint8_t attr = 0;
-					char upc[8] = {};
-					mscdex->GetUPC(drive_unit,attr,&upc[0]);
+					std::string upc = {};
+					mscdex->GetUPC(drive_unit,attr,upc);
 					mem_writeb(buffer+1,attr);
-					for (int i=0; i<7; i++) mem_writeb(buffer+2+i,upc[i]);
+					size_t upc_index = 0;
+					// Convert from ASCII to BCD encoding
+					for (int i = 0; i < 7; i++) {
```

Comment:
Extract to `ascii_to_bcd` util function in `math_utils.h`, then add some unit tests?

Reply:
Those existing functions are not suitable for this use-case.  They take in a decimal rather than a string and they're hard-coded to 4 digits.  I extracted this into a separate function with a unit test though.

Reply:
Cool, I haven't examined them too deeply, I admit.. :)